pkg/selector/pod/selector.go

```diff
@@ -146,151 +134,78 @@ func SelectAndFilterPods(ctx context.Context, c client.Client, r client.Reader,
 func SelectPods(ctx context.Context, c client.Client, r client.Reader, selector v1alpha1.PodSelectorSpec, clusterScoped bool, targetNamespace string, enableFilterNamespace bool) ([]v1.Pod, error) {
-	// TODO: refactor: make different selectors to replace if-else logics
-	var pods []v1.Pod
-
 	// pods are specifically specified
 	if len(selector.Pods) > 0 {
-		for ns, names := range selector.Pods {
-			if !clusterScoped {
-				if targetNamespace != ns {
-					log.Info("skip namespace because ns is out of scope within namespace scoped mode", "namespace", ns)
-					continue
-				}
-			}
-			for _, name := range names {
-				var pod v1.Pod
-				err := c.Get(ctx, types.NamespacedName{
-					Namespace: ns,
-					Name:      name,
-				}, &pod)
-				if err == nil {
-					pods = append(pods, pod)
-					continue
-				}
-
-				if apierrors.IsNotFound(err) {
-					log.Error(err, "Pod is not found", "namespace", ns, "pod name", name)
-					continue
-				}
-
-				return nil, err
-			}
-		}
-
-		return pods, nil
+		return selectSpecifiedPods(ctx, c, selector, clusterScoped, targetNamespace, enableFilterNamespace)
 	}
 
-	if !clusterScoped {
-		if len(selector.Namespaces) > 1 {
-			return nil, fmt.Errorf("could NOT use more than 1 namespace selector within namespace scoped mode")
-		} else if len(selector.Namespaces) == 1 {
-			if selector.Namespaces[0] != targetNamespace {
-				return nil, fmt.Errorf("could NOT list pods from out of scoped namespace: %s", selector.Namespaces[0])
-			}
-		}
+	selectorRegistry := newSelectorRegistry(ctx, c, selector)
+	selectorChain, err := registry.Parse(selectorRegistry, selector.GenericSelectorSpec, generic.Option{
+		ClusterScoped:         clusterScoped,
+		TargetNamespace:       targetNamespace,
+		EnableFilterNamespace: enableFilterNamespace,
+	})
+	if err != nil {
+		return nil, err
 	}
 
-	var listOptions = client.ListOptions{}
-	if !clusterScoped {
-		listOptions.Namespace = targetNamespace
-	}
-	if len(selector.LabelSelectors) > 0 || len(selector.ExpressionSelectors) > 0 {
-		metav1Ls := &metav1.LabelSelector{
-			MatchLabels:      selector.LabelSelectors,
-			MatchExpressions: selector.ExpressionSelectors,
-		}
-		ls, err := metav1.LabelSelectorAsSelector(metav1Ls)
-		if err != nil {
-			return nil, err
-		}
-		listOptions.LabelSelector = ls
+	var pods []v1.Pod
+	pods, err = listPods(ctx, c, r, selector, selectorChain, enableFilterNamespace)
+	if err != nil {
+		return nil, err
 	}
 
-	listFunc := c.List
-
-	if len(selector.FieldSelectors) > 0 {
-		listOptions.FieldSelector = fields.SelectorFromSet(selector.FieldSelectors)
-
-		// Since FieldSelectors need to implement index creation, Reader.List is used to get the pod list.
-		// Otherwise, just call Client.List directly, which can be obtained through cache.
-		if r != nil {
-			listFunc = r.List
+	filterPods := make([]v1.Pod, 0, len(pods))
+	for _, pod := range pods {
+		if selectorChain.Match(&pod) {
+			filterPods = append(filterPods, pod)
```

Comment:
`selectorChain` is a parameter of function `listPods`, how about moving these codes into function `listPods`?

Reply:
done